libcontainer/configs/validate/validator.go

```diff
@@ -262,16 +253,72 @@ func cgroupsCheck(config *configs.Config) error {
 	return nil
 }
 
+func checkIDMapMounts(config *configs.Config, m *configs.Mount) error {
+	if !m.IsIDMapped() {
+		return nil
+	}
+
+	if !m.IsBind() {
+		return fmt.Errorf("invalid mount %+v: GidMappings/UidMappings is supported only for mounts with the option 'bind'", m)
+	}
+	if config.RootlessEUID {
+		return fmt.Errorf("invalid mount %+v: GidMappings/UidMappings is not supported when runc is being launched with EUID != 0, needs CAP_SYS_ADMIN on the runc parent's user namespace", m)
+	}
+	if len(config.UidMappings) == 0 || len(config.GidMappings) == 0 {
+		return fmt.Errorf("invalid mount %+v: not yet supported to use GidMappings/UidMappings in a mount without also using a user namespace", m)
+	}
+	if !sameMapping(config.UidMappings, m.UIDMappings) {
```

Comment:
I read the discussions in `https://github.com/opencontainers/runtime-spec/pull/1143`.
I think this is too strict here. It will make the upstream projects can't specific the mount idmap params. I think maybe we should remove this check, let it check by kenel or upstream projects?
@thaJeztah @AkihiroSuda WDYT

Reply:
I'm not sure what part of the discussion is relevant for this.

This is strict, yes, but this shortcut simplifies the initial implementation significantly as I mention in the PR description:
> This limitation is to
simplify the initial implementation, as all our users so far only need
this, and we can avoid sending over netlink the mappings, creating a
userns with this custom mapping, etc. Future PRs will remove this
limitation.

As I said there, this is all that all the upstream projects need so far (and we don't have plans to change that). This PR is already quite big as it is, I think it makes lot of sense to have this limitation in the initial implementation and lift it in follow-up PRs.

Reply:
As you say, this PR is already quite big, I think it's quite big enough now. So I think to support idmap mount options may not make it too more bigger to review. But if you say this is all that all the upstream projects need so far, I have no other opinion.
LGTM except https://github.com/opencontainers/runc/pull/3717#discussion_r1259104260

Reply:
Cool, I'll mark this as solved, then! :)

When I played with that, what wasn't trivial is changing the netlink protocol msg and then in C forking, creating a userns, killing that process after you used the mapping, etc. So I'm happy we can tackle that in a follow-up PR :)